### PR TITLE
HEEDLS-996 add authorisation policies to bare authorised pages

### PR DESCRIPTION
**Pull request by JBsoftwire**
### JIRA link
_[HEEDLS-996](https://softwiretech.atlassian.net/browse/HEEDLS-996)_

### Description
Added appropriate policies to ApplicationSelectorController and VerifyEmailController so that they are no longer. Also involves a new policy and a refactor of the CustomPolicies to de-duplicate code.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [X] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [ ] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [X] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

**Review by simonbrent — CHANGES_REQUESTED** [review]
Additional considerations:

* In `DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml` there is a check that `User.Identity.IsAuthenticated && User.GetAdminId() != null`. If you implement my suggestions, you can replace this with `User.IsAdminAccount()`. This will ensure consistency between who can access the ApplicationSelector route and who sees a link to it.
* In `DigitalLearningSolutions.Web/Views/Home/Welcome.cshtml`, there is a link to the ApplicationSelector which is visible to any logged in user. I know that as a logged in user you only get to see this page if you have an admin account, as delegate-only accounts get redirected away to /LearningPortal/Current, but it seems wrong to me to have the view (which doesn't know anything about the redirect behaviour) logically allowing non-admins to see the link for switching applications, and only denying them that ability via action-at-a-distance. @SteveJacksonSoft do you have any thoughts on this?

**Comment by JBsoftwire** [comment]
I discussed the case of Welcome.cshtml with @SteveJacksonSoft while planning the ticket. We decided to leave it as-is because a logged-in non-admin shouldn't get to that page to begin with, and there was an edge case (I admit I forget exactly what) where it was expected behaviour.

**Review by simonbrent — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs [hunk 1]
```diff
@@ -54,10 +53,10 @@ AuthorizationPolicyBuilder policy
...
             return policy.RequireAssertion(
-                context => UserIsAdmin(context.User) &&
-                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsFrameworkDeveloper) == true) |
-                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsFrameworkContributor) == true) |
-                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsWorkforceManager) == true) |
+                context => context.User.IsAdminAccount() &&
+                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsFrameworkDeveloper) == true) ||
+                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsFrameworkContributor) == true) ||
+                           (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsWorkforceManager) == true) ||
                            (context.User.GetCustomClaimAsBool(CustomClaimTypes.IsWorkforceContributor) == true)
```
[the remark this answers is not in the document]
Comment (JBsoftwire) [reply DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs:60]: Oh of course. I was wondering why it hadn't been done with `||` to begin with if it was that simple a change!

**Review by simonbrent — APPROVED** [review]
Great, thanks

**Pull request merged**